### ISSUE TITLE
Little-CMS fast-float plugin

### ISSUE THREAD

**Issue by heckflosse** [Beep6581/RawTherapee#6016]
Little-CMS has a new fast-float plugin which speeds up transforms.

To use it, you have to build Little-CMS from scratch, which is easy:

```
git clone https://github.com/mm2/Little-CMS.git lcms2
cd lcms2
./configure --with-fastfloat
make install
```

I will create a branch for using it.

A first test saving a 45.6 Mp file full size to jpeg using RTv4_sRGB as output profile gave this:
```
without fast-float plugin: lab2rgbOut took 5312 ms
with fast-float plugin: lab2rgbOut took 754 ms
```

**Comment by Thanatomanic** [comment]
I'll close https://github.com/Beep6581/RawTherapee/issues/5794 since this is already more descriptive. I assume there is no loss in fidelity when using fast-float?

**Comment by heckflosse** [comment]
> I assume there is no loss in fidelity when using fast-float?

See this comment.

https://github.com/mm2/Little-CMS/commit/e2b6e7e06520123a9821f0381877dfe94cefb2e5#commitcomment-44770007

**Comment by heckflosse** [comment]
I created the `lcms2-fast-float-plugin` branch and added a commit which allows testing on saving files.

To compare before/after, just comment out line 379 in main.cc

**Comment by TechXavAL** [comment]
I've tried to test it, with no luck.

I'm working on a Linux Mint 20. And following @heckflosse directions:
* I have compiled lcms2 with:
  > git clone https://github.com/mm2/Little-CMS.git lcms2
  > cd lcms2
  > sudo ./configure --with-fastfloat
  > sudo make install

  I added _sudo_ at the configure part just in case... Everything seemed to go well (no errors)
* I have compiled after that the _rawtherapee-lcms2-fast-float-plugin_ branch (no errors)
* when launching RT from the command line I get:
  > ./rawtherapee: error while loading shared libraries: liblcms2_fast_float.so.1: cannot open shared object file: No such file or directory

What am I doing wrong?

**Comment by heckflosse** [comment]
@TechXavAL Thanks for reporting :+1: 

On windows it builds and runs just fine. Maybe @Floessie can help for the linux builds.

**Comment by heckflosse** [comment]
I just made a test processing 20 files of size 12Mp in queue. Settings were neutral + raw ca-correction 2 iterations + rcd demosaic + capture sharpening. Output profile RTv4_sRGB. Output format jpeg.

Without fast-float plugin: 76 seconds
With fast-float plugin: 53 seconds

**Comment by Beep6581** [comment]
For reference: this plugin is not available in any release yet, it will probably be included in LCMS-2.12.

**Comment by heckflosse** [comment]
> For reference: this plugin is not available in any release yet, it will probably be included in LCMS-2.12.

Do we need it to be in a release to try it? I dont think so...

**Comment by Floessie** [comment]
> On windows it builds and runs just fine. Maybe @Floessie can help for the linux builds.

@TechXavAL You're right, it doesn't work out of the box on Linux because library paths are cached. Although CMake includes `/usr/local` before `/usr` and builds and links just fine, `ldd` confirms that `liblcms2.so` is still taken from `/usr/lib`. So after installing LCMS with `sudo make install`, run `sudo ldconfig` to refresh the loader cache.

> I added sudo at the configure part just in case...

I wouldn't do that.

Best,
Flössie

**Comment by heckflosse** [comment]
I next step we need a cache for the transformations to avoid the ~100 ms penalty when executing `cmsCreateTransform` using the fast-float plugin. In full-size processing the penalty does not really matter, but when using fast-float plugin for monitor profile, we would get this penalty for every thumb...

**Comment by TechXavAL** [comment]
@Floessie , @heckflosse 

I haven't recompiled and installed lcms2, as I understand it was done correctly.

No I have executed `sudo ldconfig` and after that I have performed a clean compilation of the branch (I deleted both the source-code and compilation folders). Now RT runs without a hiccup.

These are my times for 100 raw files (14,5 MB each), with several tools applied, exported as TIFF 16bit uncompressed:
* without the plugin: 9:09
* with the plugin: 8:53

_create transform_ took around 34-35ms
_exec transform_ took around 120-135ms

I get around the same times exporting to jpeg, quality 95.

Well, there is an improvement, but I don't see anything to be really excited about (around a 3%). I must tell that to me any improvement is really welcomed, but looking at the data in lcms website and the times related by @heckflosse, I can clearly state that I have done something wrong.

**Comment by heckflosse** [comment]
@TechXavAL 

> create transform took around 34-35ms
exec transform took around 120-135ms

What are this times without the fast-float plugin?

**Comment by heckflosse** [comment]
@TechXavAL In my first commit, the improvement was only for queue/saveas processing, not for rawtherapee-cli...

**Comment by heckflosse** [comment]
@TechXavAL What was your output size?

**Comment by TechXavAL** [comment]
> What are this times without the fast-float plugin?

Sorry. I can't figure out how to get those values launching my dev compilation from shell. I get demosaicing, rgb and ca values, but nothing else.

> the improvement was only for queue/saveas processing, not for rawtherapee-cli

My tests have been always from GUI.

> What was your output size?

You mean the size of the tiffs? 97,0 MB. And 1,6 MB for the jpegs

Edit: ooops! I have just remebered that I have to comment one line and recompile...
I'm going to do that ASAP

**Comment by heckflosse** [comment]
> Edit: ooops! I have just remebered that I have to comment one line and recompile...
I'm going to do that ASAP

:+1:

**Comment by TechXavAL** [comment]
@heckflosse 

Line 379 in main.cc is an empty line...

Could you copy/paste the text present in the line that has to be commented, please?

**Comment by heckflosse** [comment]
It's line 1038 in iccstore.cc now

```
    cmsPlugin(cmsFastFloatExtensions());
```

**Comment by TechXavAL** [comment]
Checked.

Overal time 9:38 (a bit slower)
_create transform_ 0ms
_exec transform_ 585-645ms

So as I see it: the change of _create_ and _exec_ timings give and improvement of around 45secs for 100 images. And the most interesting thing is that with current dev the process is faster than with this branch without the plugin...

I keep thinking that my compilations and lcms2 install are faulty somehow.

**Comment by heckflosse** [comment]
@TechXavAL 
Can you check again by commenting out line 33 of iccstore.h instead of line 1038 of iccstore.cc ?
That should give the same times as dev.

For final version also the create transform will be cached, which will give a further speedup.

**Comment by TechXavAL** [comment]
:+1: 

> For final version also the create transform will be cached, which will give a further speedup.

Ok. But I can't understand why the improvement I see here are not as good as yours. I know your machine is better than mine, but still: your times are around 7x faster, but mine are just 1.08x faster. The CPU difference doesn't explain this.

Let's wait until I perform the last check

**Comment by heckflosse** [comment]
> but still: your times are around 7x faster

They are not 7 times faster when measuring a full processing:
> Without fast-float plugin: 76 seconds
With fast-float plugin: 53 seconds

When comparing full processing we need to use the same pp3 file.

That's the reason I want to know the times for create transform and especially exec transform.

**Comment by TechXavAL** [comment]
Test done.

Same timings and overall time is 9:34.

I have rechecked the same batch with RTdev and it has taken 9:13 to finish. But I don't have the timings of the transforms.

**Comment by heckflosse** [comment]
@TechXavAL 
On my machine dev and this branch without the fast float plugin have identical processing times. I just measured dev on 20 files as above and it took 76 seconds (as above measured on this branch without fast-float plugin)

**Comment by TechXavAL** [comment]
Well, it must be something in my OS.

The only thing I can think of to solve this is perform a clean OS install and then recheck everything, but I don't have an empty drive at hand, and the one I'm waiting for will arrive in about 2 weeks...

I could create a new, clean virtual system, but I don't think the timings would be useful in such case.

And just for information, this is the make part of my compilation script (most probably it doesn't make any difference, but just in case):

> cmake \
>     -DCMAKE_BUILD_TYPE="$buildType"  \
>     -DCACHE_NAME_SUFFIX="5-$master" \
>     -DPROC_TARGET_NUMBER="2" \
>     -DBUILD_BUNDLE="ON" \
>     -DBUNDLE_BASE_INSTALL_DIR="${installDir}" \
>     -DOPTION_OMP="ON" \
>     -DWITH_LTO="OFF" \
>     -DWITH_PROF="OFF" \
>     -DWITH_SAN="OFF" \
>     -DWITH_SYSTEM_KLT="OFF" \
>     -DWITH_BENCHMARK="OFF" \
>     "$HOME/${masterfolder}/${sourcefolder}-${prog}"

**Comment by heckflosse** [comment]
> Well, it must be something in my OS.

I don't think so, but atm I have no idea what causes it...
